nb2workflow/galaxy.py

```diff
@@ -0,0 +1,591 @@
+                if val == self.default_value:
+                    attrs['selected'] = 'true'
+                option = ET.SubElement(element, 'option', **attrs)
+                option.text = str(val)
+        
+        # TODO: do we need additional validation?
+        
+        return element
+
+class GalaxyOutput:
+    def __init__(self, name, is_oda, dprod=None):
+        self.name = name
+        if dprod is None:
+            dprod=''
+            self.dprod = dprod
+        else:
+            self.dprod = dprod
+            dprod += '_'
+        self.dataname = f"out_{dprod}{self.name}"
+        self.is_oda = is_oda
+        self.outfile_name = f"{name}_galaxy.output"
+    
+    @classmethod
+    def from_inspect(cls, outp_details, ontology_path, dprod=None):
+        onto = Ontology(ontology_path)
+
+        owl_uri = outp_details['owl_type']
+        if outp_details['extra_ttl'] is not None:
+            onto.parse_extra_triples(outp_details['extra_ttl'])
+        
+        if onto.is_data_product(owl_uri, include_parameter_products=False):
+            is_oda = True
+        else:
+            is_oda = False
+        
+        return cls(outp_details['name'], is_oda, dprod)
+
+    def to_xml_tree(self):
+        if self.dprod is None:
+            label = "${tool.name} -> %s"%self.name 
+        else:
+            label = "${tool.name} -> %s %s"%(self.dprod, self.name)
+            
+        attrs = {'label': label,
+                 'name': self.dataname,
+                 #'auto_format': 'true', 
+                 'format': 'auto',
+                 'from_work_dir': self.outfile_name}
+        
+        element = ET.Element('data', **attrs)
+
+        return element
+    
+        
+            
+def _nb2script(nba, ontology_path):
+    input_nb = nba.notebook_fn
+    mynb = nbformat.read(input_nb, as_version=4)
+    outputs = nba.extract_output_declarations()
+    
+    inject_import = nbformat.v4.new_code_cell(
+        dedent( """
+                import json
+                import os
+                import shutil
+                import sys
+                
+                try:
+                    import numpy as np
+                    _numpy_available = True
+                except ImportError:
+                    _numpy_available = False
+                
+                try:
+                    from oda_api.json import CustomJSONEncoder
+                except ImportError:
+                    from json import JSONEncoder as CustomJSONEncoder
+                
+                _galaxy_wd = os.getcwd()
+                """))
+    inject_import.metadata['tags'] = ['injected_import']
+    mynb.cells.insert(0, inject_import)
+    
+    inject_pos = 0
+    for pos, cell in enumerate(mynb.cells):
+        if 'parameters' in cell['metadata'].get('tags', []):
+            inject_pos = pos + 1
+            break
+    # NOTE: validation of args is external
+    inject_read = nbformat.v4.new_code_cell(
```

Comment:
am not quite sure why you add some NB cells and then export it, you could just add to export? 
but this could be useful in the future

Reply:
It just seems easier to define the right place for insertion in notebook rather than searching for it in the exported script